MAINT: switch from `numpy.array_api` to `array-api-strict`

rgommers opened the pull request:
This is a standalone version of `numpy.array_api` that will be stable, and support multiple NumPy versions (at least 1.26 and 2.0 right now).

`numpy.array_api` is likely to be removed before the numpy 2.0 release, so we should make this switch now.

rgommers committed:
MAINT: switch from `numpy.array_api` to `array-api-strict`

This is a standalone version of `numpy.array_api` that will be stable,
and support multiple NumPy versions (at least 1.26 and 2.0 right now).

`numpy.array_api` is likely to be removed before the numpy 2.0 release,
so we should make this switch now.

[skip cirrus]

lucascolley commented on scipy/fft/tests/test_helper.py:
```diff
@@ -456,8 +456,8 @@ def test_definition(self, xp):
 
-    # fft not yet implemented by numpy.array_api
-    @skip_if_array_api_backend('numpy.array_api')
+    # fft not yet implemented by array_api_strict
```
```suggestion
    # fft not yet implemented by array-api-strict
```

lucascolley commented on .github/workflows/array_api.yml:
```diff
@@ -92,6 +92,6 @@ jobs:
         # --array-api-backend == -b
-        python dev.py --no-build test -b pytorch -b numpy -b numpy.array_api -s cluster -- --durations 3 --timeout=60
-        python dev.py --no-build test -b pytorch -b numpy -b numpy.array_api -s fft -- --durations 3 --timeout=60
+        python dev.py --no-build test -b pytorch -b numpy -b array-api-strict -s cluster -- --durations 3 --timeout=60
+        python dev.py --no-build test -b pytorch -b numpy -b array-api-strict -s fft -- --durations 3 --timeout=60
```
```suggestion
        python dev.py --no-build test -b pytorch -b numpy -b array_api_strict -s cluster -- --durations 3 --timeout=60
        python dev.py --no-build test -b pytorch -b numpy -b array_api_strict -s fft -- --durations 3 --timeout=60
```

lucascolley reviewed (changes requested):
some complaints from Mypy. We can also remove the reference to `numpy.array_api` in `pytest.ini`.

rgommers commented:
Ah, failures in the array API CI job because the FFT support got more strict on supported dtypes:
```
TypeError: Only complex floating-point dtypes are allowed in fft
...
TypeError: Only floating-point dtypes are allowed in fftshift
```

rgommers committed:
TST: fft: ensure we use floating-point dtypes as needed

Integer dtypes are rejected by array-api-strict; the relevant
functions here only support floating-point input, tests using
integers that then get cast under the hood is a shortcut that
didn't matter much till now.

[skip cirrus]

lucascolley commented on environment.yml:
```diff
@@ -3,7 +3,7 @@
 #   $ conda activate scipy-dev
 #
-name: scipy-dev
+name: scipy-dev-tmp4
```
```suggestion
name: scipy-dev
```

lucascolley commented on scipy/fft/tests/test_basic.py:
```diff
@@ -114,11 +133,12 @@ def test_ifftn(self, xp):
             for norm in [None, "backward", "ortho", "forward"]:
                 xp_assert_close(fft.rfft(x, n=n, norm=norm),
-                                fft.fft(x, n=n, norm=norm)[:(n//2 + 1)])
+                                fft.fft(xp.asarray(x, dtype=xp.complex128),
```
```suggestion
                                fft.fft(xp.astype(x, dtype=xp.complex128),
```

rgommers replied:
good point, we've used `astype` in other places already so should work fine and be more accurate

lucascolley commented on scipy/fft/tests/test_basic.py:
```diff
@@ -135,8 +155,8 @@ def test_irfft(self, xp):
-        x = xp.asarray(random((30, 20)))
-        expect = fft.fft2(x)[:, :11]
+        x = xp.asarray(random((30, 20)), dtype=xp.float64)
+        expect = fft.fft2(xp.asarray(x, dtype=xp.complex128))[:, :11]
```
```suggestion
        expect = fft.fft2(xp.astype(x, dtype=xp.complex128))[:, :11]
```

lucascolley commented on scipy/fft/tests/test_basic.py:
```diff
@@ -156,8 +176,8 @@ def test_irfft2(self, xp):
-        x = xp.asarray(random((30, 20, 10)))
-        expect = fft.fftn(x)[:, :, :6]
+        x = xp.asarray(random((30, 20, 10)), dtype=xp.float64)
+        expect = fft.fftn(xp.asarray(x, dtype=xp.complex128))[:, :, :6]
```
```suggestion
        expect = fft.fftn(xp.astype(x, dtype=xp.complex128))[:, :, :6]
```

lucascolley commented on scipy/fft/tests/test_basic.py:
```diff
@@ -211,8 +231,8 @@ def test_hfft2(self, xp):
-        x = xp.asarray(random((30, 20)))
-        expect = fft.ifft2(x)[:, :11]
+        x = xp.asarray(random((30, 20)), dtype=xp.float64)
+        expect = fft.ifft2(xp.asarray(x, dtype=xp.complex128))[:, :11]
```
```suggestion
        expect = fft.ifft2(xp.astype(x, dtype=xp.complex128))[:, :11]
```

lucascolley commented on scipy/fft/tests/test_basic.py:
```diff
@@ -234,8 +254,8 @@ def test_hfftn(self, xp):
-        x = xp.asarray(random((30, 20, 10)))
-        expect = fft.ifftn(x)[:, :, :6]
+        x = xp.asarray(random((30, 20, 10)), dtype=xp.float64)
+        expect = fft.ifftn(xp.asarray(x, dtype=xp.complex128))[:, :, :6]
```
```suggestion
        expect = fft.ifftn(xp.astype(x, dtype=xp.complex128))[:, :, :6]
```

lucascolley commented on scipy/fft/tests/test_basic.py:
```diff
@@ -490,8 +526,13 @@ def test_not_last_axis_success(self, xp):
+    if func in [fft.rfft, fft.rfftn, fft.ihfft]:
+        dtype = xp.float64
+    else:
+        dtype=xp.complex128
```
```suggestion
        dtype = xp.complex128
```

lucascolley commented on scipy/fft/tests/test_basic.py:
```diff
@@ -346,20 +373,29 @@ def test_dtypes_nonstandard(self, dtype):
-    def test_dtypes(self, dtype, xp):
+    def test_dtypes_real(self, dtype, xp):
         x = xp.asarray(random(30), dtype=getattr(xp, dtype))
         out_dtypes = {"float32": xp.complex64, "float64": xp.complex128}
         x_complex = xp.asarray(x, dtype=out_dtypes[dtype])
```
```suggestion
```

rgommers commented on scipy/fft/tests/test_basic.py:
```diff
@@ -346,20 +373,29 @@ def test_dtypes_nonstandard(self, dtype):
+    @pytest.mark.parametrize("dtype", ["complex64", "complex128"])
+    def test_dtypes_complex(self, dtype, xp):
+        x = xp.asarray(random(30), dtype=getattr(xp, dtype))
+        out_dtypes = {"complex64": xp.complex64, "complex128": xp.complex128}
+        x_complex = xp.asarray(x, dtype=out_dtypes[dtype])
```
```suggestion
```

rgommers commented on scipy/fft/tests/test_basic.py:
```diff
@@ -346,20 +373,29 @@ def test_dtypes_nonstandard(self, dtype):
+        res_fft = fft.ifft(fft.fft(x))
+        # Check both numerical results and exact dtype matches
+        rtol = {"complex64": 1.2e-4, "complex128": 1e-8}[dtype]
+        xp_assert_close(res_fft, x_complex, rtol=rtol, atol=0)
```
```suggestion
        xp_assert_close(res_fft, x, rtol=rtol, atol=0)
```

rgommers commented:
```
FAILED scipy/fft/tests/test_basic.py::TestFFT1D::test_rfft[numpy] - AttributeError: module 'numpy' has no attribute 'astype'
```
that is why I avoided it in test code I think. In the library code, I think we always ensure to use the compat library, but not in tests.

lucascolley commented:
Ah, sorry, I remember now, makes sense!

rgommers commented:
Seems all happy now, thanks for the quick review @lucascolley. @tylerjereddy you may want to have a look as well, since this is a dependency change and may intersect with something you're working on.

lucascolley reviewed (approved):
LGTM

tylerjereddy reviewed (approved):
The dtype handling changes around FFT aren't exactly pretty, but sounds like you and Lucas are happy with that. I checked the diff and seems "ok" to me. I also checked the testsuite totals in a few situations with `cuda` device (`SCIPY_DEVICE=cuda python dev.py test -j 32 -b all`):

- `main`: `6 failed, 47829 passed, 5690 skipped, 149 xfailed, 12 xpassed in 54.73s`
- here without the new package: `6 failed, 46148 passed, 5684 skipped, 149 xfailed, 12 xpassed in 55.22s`
- here with the new package: `6 failed, 47837 passed, 5690 skipped, 149 xfailed, 12 xpassed in 51.00s`

That seems "ok" to me--some of the skipping arithmetic isn't correct yet, but that's because I need to get Lucas' other test machinery PR in.

My WIP array API stuff has progressed too slowly to justify blocking anything else, so in it goes.

rgommers commented:
> there'll be a small merge conflict on my PR now, will get to it soon

thanks Lucas. it's really time to get that one merged, I'll make it a priority
